design-proposals/klip-38-variable-substitution.md

```diff
@@ -0,0 +1,302 @@
+SET VAR;       # lists all variables
+SET;           # what to do? lists all server, cli, variables, etc?
+```
+
+Also, the use of single-quotes in `SET` server settings is tedious to type. So after removing them, we'll have two different sets:
+* `SET 'server_property' = 'server_value'`
+* `SET VAR variable = `variable_value'
+
+This looks confusing too. So we have to decide for a new syntax for substitution variables.
+
+Other DBs use different syntax to create user variables (DECLARE, DEFINE, \set, SET). For ksqlDB, I'll choose to use `DEFINE` and `UNDEFINE` which
+is what Oracle SQL*plus uses. Any user coming from this DB will immediately understand the command.  
+This is opened to discussion, but will continue with this for now.
+
+#### Creating and printing substitution variables
+
+The command to interact with substitution variables is `DEFINE`.  
+`DEFINE` will allow us to define a new variable, print the variable value and list all current defined variables.
+
+Syntax:
+```
+DEFINE [name [= 'value']];
+
+Where: 
+  name     is the variable name to define
+  value    is the variable value
+```
+Valid variables names start with a letter or underscore (\_) followed by zero or more alphanumeric characters or underscores.
+
+All variable values need to be wrapped using single quotes. All variables are stored as string values in memory, and when
+replaced, the value without the quotes will be used. ksqlDB parsing will be in charge of determining the format at that
+time.
+
+The reason not to remove single-quotes from values is to allow spaces, and to leave the syntax open for using data types in the future if required.  
+Also, having decide the format at run-time (i.e. `define n = 3.34238`) is tricky as we could lose precision in case of decimals. Better stay with Strings for now.
+
+Define a new variable:
+```
+ksql> DEFINE replicas = '3';
+ksql> DEFINE partitions = '5';
+```
+
+Print a defined variable value:
+```
+ksql> DEFINE replicas;
+DEFINE replicas   = '3'
+```
+
+List all defined variables:
+```
+ksql> DEFINE;
+DEFINE replicas   = '3'
+DEFINE partitions = '5'
+```
+
+#### Deleting substitution variables
+
+The `UNDEFINE` command deletes the variable.
+
+Syntax:
+```
+UNDEFINE name;
+```
+
+### 2. Scope
+
+This variable has a session scope.
+
+Variables will live during the session of a ksqlDB connection, then destroyed when the connection is closed. Only the user
+in that session can interact with the defined substitution variables. No other users can interact with other users' variables.
+
+In CLI, the session is opened when starting the CLI with the `ksql` command, and destroyed when typing `exit`.
+
+In Java API client, the session is opened when `Client.create()` is called, and closed when `Client.close()` is called. Variables
+may be defined using a client method instead of a new syntax. This is to ensure users know the scope starts after the `create()` call.
+
+TBD: Define a Java API method for variable substitution.
```

Comment:
+1 to defining a new method for this rather than trying to piggyback off existing methods. If we don't support variable substitution in the REST API, the Java client will have to maintain variables in its own code, similar to the CLI. If we updated the CLI to use the Java client (https://github.com/confluentinc/ksql/issues/6269) we'd have fewer issues like these in the future but that's out of scope for this work for sure.